### PR TITLE
Fix for namespace not being propagated on Auth schemes (except Token)

### PR DESCRIPTION
**Pull request by rosuvlad**
Add X-Vault-Namespace also to auth requests. 
Would get 400 Bad Request - Invalid Role ID (for AppRole auth) otherwise.

### CODE REVIEW

**Comment by rosuvlad** [comment]
Using:

springCloudVersion: Hoxton.SR1
org.springframework.boot: 2.2.4.RELEASE
org.springframework.cloud:spring-cloud-config-server: 2.2.1.RELEASE
We are having Enterprise Vault available in our company and we need to use the namespace segregation.

The problem is that the namespace is enforced for Auth endpoints (except the root one, token) in our Vault configuration.

In our case AppRole would be something specific for each namespace and the current state of spring-config-server is not passing X-Vault-Namespace in the calls for the AppRole flow.
By running the code without the fix, we would encounter a 400 - BadRequest - Invalid Role ID from the Vault backend.

**Comment by rosuvlad** [comment]
Original wrongfully targeted PR: https://github.com/spring-cloud/spring-cloud-config/pull/1564

**Comment by spencergibb** [comment]
@scottfrederick or @mp911de?

**Comment by kamalakarp** [comment]
@scottfrederick @mp911de - Can you please let us know if we need to do anything to push this PR ?

**Comment by kamalakarp** [comment]
I was referring the same issue in https://github.com/spring-cloud/spring-cloud-config/issues/1589

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java [hunk 1]
```diff
@@ -125,6 +123,13 @@ public SslConfiguration sslConfiguration() {
 		return new SslConfiguration(keyStoreConfiguration, trustStoreConfiguration);
 	}
 
+	@Override
+	public RestOperations restOperations() {
```
Comment (mp911de) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java]: Note: The need to override this method should be considered a workaround. Can be removed once spring-projects/spring-vault#546 is resolved.

**Comment by mp911de** [comment]
Looks good from my perspective. The PR even revealed an inconsistency in Spring Vault.

**Comment by scottfrederick** [comment]
This change looks good to me, but there should be a tested added such that the test would have failed before this change but succeeds after the change.

**Comment by kamalakarp** [comment]
It requires enterprise vault which is licensed version and namespaces option not available in open source vault.

**Comment by mp911de** [comment]
Enterprise vault can run up to 30 minutes without a valid license to test namespaces. However, a namespace test arrangement requires a bit of CI and test fixture setup, see https://github.com/spring-projects/spring-vault/blob/master/spring-vault-core/src/test/java/org/springframework/vault/core/VaultNamespaceSecretIntegrationTests.java#L84-L125

**Comment by kamalakarp** [comment]
@rosuvlad - is this something you want to do on your fork  ?

**Comment by rosuvlad** [comment]
I can give it a try to create some tests to cover this but it will take some time. I am not proficient with this pipeline.

**Comment by scottfrederick** [comment]
@rosuvlad I can work on adding a test when the PR is merged if you'd prefer. Just let me know what you'd like to do.

**Comment by rosuvlad** [comment]
Thank you Scott! I would really like that!

**Comment by scottfrederick** [comment]
@rosuvlad This is merged with [one small change](https://github.com/spring-cloud/spring-cloud-config/commit/959f492ddf77cbfe84dc4d3f6633ebf4118019d7#diff-b1d3f3a0748166d6a2b94ac7db493ed6R133-R143). Instead of overriding `AbstractVaultConfiguration#restOperations` and explicitly adding the namespace interceptor there, this change overrides the method with an exact copy of the change that @mp911de [applied to Spring Vault](https://github.com/spring-projects/spring-vault/commit/e1d8c51999efd1a4bbb87394364a7ea3c58a888f#diff-7068d8b46418a1a6cd18c84dbe473bac). With the new tests in place, this will make it easy for us to delete this method override later when Spring Cloud Config pulls in a Spring Vault upgrade.

**Comment by rosuvlad** [comment]
Cool! Thank you for the support offered! Have a great day ahead!

**Comment by kamalakarp** [comment]
Thank you @scottfrederick @rosuvlad for the support.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java (hunk 1 above)
Comment (higonetr) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/vault/SpringVaultClientConfiguration.java:127]: Hi,
I need to implement it with namespaces. There is a better way to do it? 

`@Override
public RestOperations restOperations() {
    RestTemplate restTemplate = (RestTemplate) super.restOperations();
    // To define a namespace
    List<ClientHttpRequestInterceptor> clientHttpRequestInterceptorList = new LinkedList<>();
    ClientHttpRequestInterceptor namespaceInterceptor = VaultClients.createNamespaceInterceptor(NAMESPACE);
    clientHttpRequestInterceptorList.add(namespaceInterceptor);
    restTemplate.setInterceptors(clientHttpRequestInterceptorList);
    //

    return restTemplate;
}`

**Pull request merged**